scheduler/pkg/envoy/xdscache/seldoncache.go

```diff
@@ -352,27 +323,37 @@ func (xds *SeldonXDSCache) RemoveRoute(routeName string) error {
 	}
 	delete(xds.Routes, routeName)
 	for _, cluster := range route.Clusters {
-		err := xds.removeRouteFromCluster(routeName, route, cluster)
+		err := xds.removeRouteFromCluster(route, cluster)
 		if err != nil {
 			return err
 		}
 	}
-	for _, mirror := range route.Mirrors {
-		err := xds.removeRouteFromCluster(routeName, route, mirror)
+	if route.Mirror != nil {
+		err := xds.removeRouteFromCluster(route, *route.Mirror)
 		if err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
-func (xds *SeldonXDSCache) AddEndpoint(clusterName, upstreamHost string, upstreamPort uint32) {
-	cluster := xds.Clusters[clusterName]
-	k := fmt.Sprintf("%s:%d", upstreamHost, upstreamPort)
-	cluster.Endpoints[k] = resources.Endpoint{
-		UpstreamHost: upstreamHost,
-		UpstreamPort: upstreamPort,
+func (xds *SeldonXDSCache) removeRouteFromCluster(route resources.Route, cluster resources.TrafficSplit) error {
+	httpCluster, ok := xds.Clusters[cluster.HttpCluster]
+	if !ok {
+		return fmt.Errorf("can't find http cluster for route %s cluster %s route %+v", route.RouteName, cluster.HttpCluster, route)
+	}
+	delete(httpCluster.Routes, resources.RouteVersionKey{RouteName: route.RouteName, ModelName: cluster.ModelName, Version: cluster.ModelVersion})
+	if len(httpCluster.Routes) == 0 {
+		delete(xds.Clusters, cluster.HttpCluster)
 	}
 
-	xds.Clusters[clusterName] = cluster
+	grpcCluster, ok := xds.Clusters[cluster.GrpcCluster]
+	if !ok {
+		return fmt.Errorf("can't find grpc cluster for route %s cluster %s route %+v", route.RouteName, cluster.GrpcCluster, route)
+	}
+	delete(grpcCluster.Routes, resources.RouteVersionKey{RouteName: route.RouteName, ModelName: cluster.ModelName, Version: cluster.ModelVersion})
+	if len(grpcCluster.Routes) == 0 {
+		delete(xds.Clusters, cluster.GrpcCluster)
+	}
```

Comment:
can this be refactor into a helper used by http and grpc?

Reply:
Yes.